libc/src/math/generic/asinhf16.cpp

```diff
@@ -0,0 +1,84 @@
+//===-- Half-precision asinhf16(x) function--------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception.
+//
+//
+//===----------------------------------------------------------------------===//
+
+#include "src/math/asinhf16.h"
+#include "src/__support/FPUtil/except_value_utils.h"
+#include "src/__support/FPUtil/generic/sqrt.h"
+#include "src/__support/FPUtil/multiply_add.h"
+#include "src/__support/common.h"
+#include "src/__support/macros/config.h"
+#include "src/__support/macros/properties/types.h"
+#include "src/math/generic/explogxf.h"
+
+namespace LIBC_NAMESPACE_DECL {
+
+#ifndef LIBC_MATH_HAS_SKIP_ACCURATE_PASS
+static constexpr size_t N_EXCEPTS = 8;
+
+static constexpr fputil::ExceptValues<float16, N_EXCEPTS> ASINHF16_EXCEPTS{
+    {// (input, RZ output, RU offset, RD offset, RN offset)
+     {0x3769, 0x372A, 1, 0, 1},
+     {0x3B5B, 0x3A96, 1, 0, 0},
+     {0x4B1F, 0x42B3, 1, 0, 0},
+     {0x4C9B, 0x4336, 1, 0, 1},
+     {0xB769, 0xB72A, 0, 1, 1},
+     {0xBB5B, 0xBA96, 0, 1, 0},
+     {0xCB1F, 0xC2B3, 0, 1, 0},
+     {0xCC9B, 0xC336, 0, 1, 1}}};
+#endif // !LIBC_MATH_HAS_SKIP_ACCURATE_PASS
+
+LLVM_LIBC_FUNCTION(float16, asinhf16, (float16 x)) {
+  using FPBits = fputil::FPBits<float16>;
+  FPBits xbits(x);
+
+  float x_d = x;
+  uint16_t x_u = xbits.uintval();
+  uint16_t x_abs = x_u & 0x7fff;
+
+  if (LIBC_UNLIKELY(xbits.is_inf_or_nan())) {
+    if (xbits.is_signaling_nan()) {
+      fputil::raise_except_if_required(FE_INVALID);
+      return FPBits::quiet_nan().get_val();
+    }
+
+    return x;
+  }
+
+#ifndef LIBC_MATH_HAS_SKIP_ACCURATE_PASS
+  // Handle exceptional values
+  if (auto r = ASINHF16_EXCEPTS.lookup(x_u); LIBC_UNLIKELY(r.has_value()))
+    return r.value();
+#endif // !LIBC_MATH_HAS_SKIP_ACCURATE_PASS
+
+  const float SIGN[2] = {1.0f, -1.0f};
+  float x_sign = SIGN[x_u >> 15];
+
+  // |x| <= 0.25
+  if (LIBC_UNLIKELY(x_abs <= 0x3400)) {
+    if (LIBC_UNLIKELY(x_abs == 0))
+      return x;
+    if (LIBC_UNLIKELY((fputil::get_round() == FE_UPWARD) && (x_u >= 0x8401) &&
+                      (x_u <= 0x90E6)))
+      return static_cast<float16>(x_d + 0x1p-24f);
+
+    float x_sq = x_d * x_d;
+    // Generated by Sollya with:
+    // > P = fpminimax(asinh(x)/x, [|0, 2, 4, 6, 8|], [|SG...|],[0, 2^-2]);
+    float p = fputil::polyeval(x_sq, 1.0f, -0x1.555556p-3f, 0x1.3334dep-4f,
+                               -0x1.6f3e2p-5f, 0x1.51d012p-5f);
```

Comment:
Nit:

```suggestion
    // Generated by Sollya with:
    // > P = fpminimax(asinh(x)/x, [|0, 2, 4, 6, 8|], [|SG...|], [0, 2^-2]);
    float p = fputil::polyeval(x_sq, 1.0f, -0x1.555556p-3f, 0x1.3334dep-4f,
                               -0x1.6f3e2p-5f, 0x1.51d012p-5f);
```

Also, I don't get the same polynomial with Sollya version/commit `55574d1d`:

```
> display = hexadecimal;
Display mode is hexadecimal numbers.
> fpminimax(asinh(x)/x, [|0, 2, 4, 6, 8|], [|SG...|],[0, 2^-2]);
0x1p0 + x^0x1p1 * (-0x1.555552p-3 + x^0x1p1 * (0x1.332f6ap-4 + x^0x1p1 * (-0x1.6c53dep-5 + x^0x1p1 * 0x1.bd114ep-6)))
>
```

Reply:
In all honestly, I'm note sure how I got those coefficients. I'm fairly certain I ran that same command (without display = hexadecimal), and can't figure out how any rounding errors could have happened. But somehow the one I have currently actually performs a little better. It handles an extra edge case `0x1.af4p-3` that the actual polynomial doesn't.

I experimented a little and found that using the correct Sollya polynomial except replacing `0x1.bd114ep-6` with `0x1.bd114ep-5` also handles that extra edge case. I'd understand if you prefer we used the actual polynomial and handled that edge case separately since the change I proposed seems pretty arbitrary, but I'd like to know anyway if I should use the actual Sollya polynomial or the Sollya polynomial with that one change.

Reply:
@lntue ping

Reply:
I talked with Tue and using the Sollya polynomial with the one change to the last coefficient for better accuracy is fine as long as it's explained in the comments, which it already is.